### ISSUE TITLE
Support parallel downloads

### ISSUE THREAD

**Issue by lukechampine** [lukechampine/us#24]
Currently all `PseudoFS` methods use coarse-grained locking: they hold a mutex for the entire duration of the method, meaning you can't call `Read` in parallel with `Write`, or even with another `Read`. I designed it this way for simplicity, but now the time has come to switch to more fine-grained locking in order to increase performance.

The main contended resource here is `Session`s. Concurrent `Read` and `Write` calls must "compete" for the available `Session`s, and must block while waiting to acquire each `Session` they need. A `Write` call, if it flushes sectors, will need to access every `Session` in the `HostSet`, whereas a `Read` call may only need to access `minHosts` `Session`s. This means that, when downloading, it is possible to increase throughput by a factor of the file's redundancy. If a file is stored at 3x redundancy, you can download it 3x faster than if it was stored at 1x redundancy.

No API changes are necessary to support parallelism. Callers can simply start making `Read` and `Write` calls in goroutines, just as they can with `*os.File`. (Technically, they will need to use `ReadAt` and `WriteAt`, since `Read` and `Write` are stateful.) Example code:
```go
// split one large read into two parallel reads
buf := make([]byte, readSize)
buf1, buf2 := buf[:readSize/2], buf[readSize/2:]
var wg sync.WaitGroup
wg.Add(2)
go func() { pf.ReadAt(buf1, 0); wg.Done() }()
go func() { pf.ReadAt(buf2, readSize/2); wg.Done() }()
wg.Wait()
```
This is a little unwieldy (and doesn't handle errors), but it would be easy to add a helper function that automatically splits up work for you and handles errors properly. I think this is what https://github.com/klauspost/readahead does, so perhaps callers could just use that package.

Writes, on the other hand, cannot be significantly sped up by adding parallelism. This is because, as previously mentioned, writing requires accessing *all* of the hosts, so you'll always be bottlenecked by the slowest host. (Imagine that you have two hosts: one that takes 1s per sector, and one that takes 10s. You might be able to upload all of the first host's data before the second has finished a single sector, but this doesn't improve your overall throughput; you still need to wait for the second host to finish.) However, there are two scenarios where the parallelism does help. First, if your slowest host is only *temporarily* slow, and then greatly speeds up soon after, then parallelism allows you to "paper over" the period of temporary slowness. Second, if you are simultaneously uploading and downloading, then parallelism allows you to start downloading from a host as soon as you finish uploading to it, rather than having to wait for all hosts to finish uploading.

**Comment by vargrant** [comment]
I would like to comment this: "Writes, on the other hand, cannot be significantly sped up by adding parallelism. This is because, as previously mentioned, writing requires accessing all of the hosts, so you'll always be bottlenecked by the slowest host. (Imagine that you have two hosts: one that takes 1s per sector, and one that takes 10s. You might be able to upload all of the first host's data before the second has finished a single sector, but this doesn't improve your overall throughput; you still need to wait for the second host to finish.)"

At first you need to add some levels of abstraction for polsibiility to implement more advanced algorithm. I propose to have next entities:
1) job for upload - total data which must be uploaded
2) chunk - part of very small data from job (may be it's not need)
3) worker - it's process for upload data to host. (you will have several workers)
4) pool of available hosts
5) some datbase for posibillity to store map of allocation

Algorithm
1) You have 1000 TB of data and need to upload it's to 2 hosts. It's mean that one host must to have not more than 512 GB of data, but really all this data can be uploaded to the 3 or more hosts. May be posible to upload 724 GB to one host and 300GB to the other host.  All this must be customized by users - becuase it's deppend from the politic of security police.
2) When job is started - some proces must provide data by chunks for each worker. Worker it's also separate process - you have 2 workers for 2 hosts.
3) In the moment when worker requesting the next chunk to upload  you need to check performance by some stored map of requested chunks for this job. Here you can detect that this worker already uploaded more data then other worker and you will have possibility to provide for worker not only next chunk but also nominate for uploding and some host from your pool.

This algo is very flexible and will improve uploading speed even if you will have some slow hosts.

**Comment by lukechampine** [comment]
Parallel downloading implemented in: fb4bc71cb4ed44845dd1a6cf9ccb5c289970c697 b9a4fc28464a6838155458a63bab6d22426d3fcf c004283c84732fc1a4abc62cee9ba6198c32926c

However, the algorithm could definitely be improved. Right now, two `ReadAt` calls can contend for the same host; instead, each call should try to use hosts that are idle, and only fallback to blocking on `acquire` if necessary.

**Comment by lukechampine** [comment]
Closing this; we can address other forms of I/O parallelism in separate issues.
